### DISCUSSION TITLE
Question about Resources & the file transfer example

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#313] [Help & Questions] [answered]
Now that I understand Links (not comprehensively but good for now), I need to learn about using Resource objects so I'm looking at the file transfer example.

I figured out from the RNS source code how a resource transfer is started:
>       Advertise the resource. If the other end of the link accepts
>       the resource advertisement it will begin transferring.

Q1: the default value for advertise is True, so does the transfer start upon immediately after the Resource is created?

Most of the code for the example is pretty clear. I see how the "file_resource" (an instance of RNS.Resource) is created in the client_request handler of the server and how it's associated with the link, and sets a callback, but I don't see a "filename" property in the API or the RNS source. 

Q2: Can anyone explain this line of code in the example: file_resource.filename = filename? Should it be `file_resource.file = filename` or `file_resource.input_file = filename`? Those _are_ defined in RNS.Resource.py but I don't see filename in that source at all.

**Comment by markqvist** [comment, marked as answer]
> Q1: the default value for advertise is True, so does the transfer start upon immediately after the Resource is created?

By default, a `Resource` is advertised over the link as soon as it is created. This behaviour can be disabled if you want to do something else first, and then manually advertise it.

Once the `Resource` has been advertised, it is up to the receiver whether or not to accept it. This follows a general principle in Reticulum. It is always up to the *receiving party* to ask for more information, and stuff will not be sent to a receiver, unless it has asked for it.

> Q2: Can anyone explain this line of code in the example: file_resource.filename = filename?

The `filename` property that is set on the `Resource` instance has nothing to do with the API as such. It is just an extra property that the file-transfer program sets on the `Resource` instance to (locally) keep track of what filename it has. It's used later down in the code to print a message when the file transfer concludes.

**Reply by ThomasFreedman** [reply]
>The filename property that is set on the Resource instance has nothing to do with the API as such. It is just an extra property...

Learn something new about Python every day it seems.  I still don't understand why that line isn't flagged as an "unresolved reference", since that property is not declared in the object definition. I suspect that due to the property being added to only a local instance of the Resource object. If so, an object instance of any type can have properties added to it. A quick check shows that isn't true tho, at least in the general case:

```
strObj = "this is a string"
strObj.newProp = "test"

```
fails, no such attribute as 'newProp'

**Reply by markqvist** [reply]
Because `str` is an immutable builtin data type, that has to act nicely woth various compiled C functions, and not an instance of a class derived from a Python standard object.

In general you can do it, but not on strings.
